### DISCUSSION TITLE
Need to figure out how to sum up the updates

### DISCUSSION THREAD

**Discussion by aljawaid** [aljawaid/PluginManager#42] [Ideas]
A few times it has been requested to have a summary of some sort of available updates for installed plugins.  I must have tried about 10 times but I can't get it to work.

- Installed plugins page should show 'update' button to directly install updated version
- Plugins directory should show a total count of updates available at the top of the page

**Comment by alfredbuehler** [comment]
>     * Installed plugins page should show 'update' button to directly install updated version

Yes for the button. But each plugin needs IMHO a checkbox, e.g., if I want to upgrade all but one.

**Reply by aljawaid** [reply]
Naaa.... up to you if you want to create that as I wouldnt know how to do it.... but as a user I would never do that because I wouldnt know which plugin caused an immediate issue without interrogating the logs etc... but hey I dunno

**Reply by alfredbuehler** [reply]
You mean a button for each installed plugin with available update?

**Reply by aljawaid** [reply]
yes but the update button should be in the installed plugins.... it already exists in the directory

**Reply by alfredbuehler** [reply]
@aljawaid I'm finished, somehow. I did the following:
- Show a conditional update button in the list of installed plugins. This button is currently non-functional, I just copy/pasted to show something. As usual, all styling is up to you.
![image](https://github.com/aljawaid/PluginManager/assets/2079289/ccdf59c0-549d-4384-9891-f857a49ba2b3)

- Show the number of available updates on top of the plugin directory list. Here, the same applies: Please finish the styling.
![image](https://github.com/aljawaid/PluginManager/assets/2079289/ab907568-032d-475b-87ba-61b1b5cbaf7e)

Should I commit this in a branch for reviewing, or can I put it directly into master?

**Reply by aljawaid** [reply]
Oooh looks good. Push it directly to master. I will work on the styling later.

**Reply by aljawaid** [reply]
I think this would work for most use cases. Thanks for that, I'm looking forward to working on the styling.

**Reply by alfredbuehler** [reply]
Done.
0ecf47e3f0dd06901dc75d70fd2e082e44c85a21
eee7b3ed9d553e705a0ebb874de6c8f20e794ca5
c6d2fa8cb60471d2f9e8cea52a131460cb076471

**Reply by alfredbuehler** [reply]
Arghh... I assume I'm not responsible for that newly added code checking. AFICS, the complaints don't come from my inserted code.

**Reply by aljawaid** [reply]
hehe I will have a look, don't worry about that code checking.... I will look into it, I a learning as I go along.

**Reply by alfredbuehler** [reply]
Consider disabling the checking for templates, at least until they are cleaned up.

**Reply by aljawaid** [reply]
its fine, I only action the errors which are minimal... the warnings are set for line lengths.. slowly over time when I move the svg code out of the file the warnings will go... it stays as a reminder for me.

**Reply by aljawaid** [reply]
Show the count on the installed plugins page too https://github.com/aljawaid/PluginManager/commit/9ca1e909013336a0a88982e4d3389257e2fa924a

**Reply by aljawaid** [reply]
@alfredbuehler 

> I'm finished, somehow. I did the following:
> 
>     * Show a conditional update button in the list of installed plugins. This button is currently non-functional, I just copy/pasted to show something. As usual, all styling is up to you.

For the update button, how can I make this work?

I'm guessing I need to fix `urlencode($plugin['download']` somehow


```php
<td class="plugin-uninstall">
    <?php if ($is_configured && in_array($plugin->getPluginName(), $updatables, true)): ?>
        <button class="btn-uninstall" title="<?= t('This feature is currently unavailable') ?>">
            <?= $this->url->icon('refresh', t('Update'), 'PluginController', 'update', array('archive_url' => urlencode($plugin['download'])), true, 'update-plugin', t('Update this plugin')) ?>
        </button>
    <?php endif ?>
</td>
```
![image](https://github.com/aljawaid/PluginManager/assets/10233708/d808edf2-4269-47bf-b664-2bc1c4a789b6)

**Reply by alfredbuehler** [reply]
<rant>
It's a very complicated crap. The show.php and directory.php are using different sets of plugin information. I don't know whether this is by (IMHO poor) design or simply a result of ongoing patchwork.
This needs some time to be fixed, as I won't make it crappier as it is now. Currently, I don't have a clear vision.
</rant>

**Reply by aljawaid** [reply]
I totally agree, the core is a mess for plugins.... I will just change the button to a link to the directory for now

**Reply by alfredbuehler** [reply]
I had an Idea, will commit soon. Seems to work, but I've simply stolen your update button from directory.php.
It needs your styling, please. :wink: 
![image](https://github.com/aljawaid/PluginManager/assets/2079289/d8a9d00c-f787-45ff-9a1f-5ddec2782e66)

**Reply by aljawaid** [reply]
yeh sure, i will wait for your commit

**Reply by alfredbuehler** [reply]
Argh, there is still an issue. It updates the plugin, but that update function doesn't reload the page, means the update button remains until reload.
Should I commit anyway?

**Reply by aljawaid** [reply]
sorry I had fallen asleep, thanks for the commit, I'm just working on the button styling

**Reply by alfredbuehler** [reply]
That reload thing is still pending. Currently, I don't have any solution.

**Reply by aljawaid** [reply]
> That reload thing is still pending. Currently, I don't have any solution.

I'm just wondering, should we make it into a confirm modal? Then maybe the reload will work?  Or maybe duplicate the controller action into the plugin and add the reload there.

Just a few thoughts

**Reply by alfredbuehler** [reply]
Maybe it's worth a try. But it's up to you, I don't know how to use it.

> Or maybe duplicate the controller action into the plugin and add the reload there.

Won't work either, as we remain on the same page, which is still valid, in browser speaking.
The only solution would be to invalidate the page so that the browser is forced to reload. But for this, the request must be changed, AFAIK.
I'm in lack of experience in that area.

**Reply by aljawaid** [reply]
ok I have just finished the styling of it... I will think about what you said.... maybe @creecros has any ideas? (the button works and updates the plugin, but doesnt reload the page).

@alfredbuehler thanks for your help in getting the functioning working at least. We're 99% there :)

**Reply by aljawaid** [reply]
So I tried a few things, including wrapping the `button` into a `form` and also using `js` `onclick` function but nothing worked to reload the page.

For the next release, as the function works itself and the button does update the plugin, I have just put a notice to remind users to refresh the page.

![image](https://github.com/aljawaid/PluginManager/assets/10233708/d5ef6004-fe7b-4d39-9186-3de8c3e3cf50)

**Reply by aljawaid** [reply]
I'm starting to think thi is normal behaviour (even if wrong).  I just updated a plugin from the directory page, sure it installed and redirected to installed plugins...but it never updated the plugin details until i clicked the page through the side menu.

The best way to check is with template manager.  The old code used the word 'update' in the plugin description.  Now i updated through the directory and got redirected.  The page should have had the content refreshed also. But when I click on updates in teh search filter, the old plugin description got included.  If i went through the page again through the sidebar, the problem was resolved.

Some sort of caching issue with the core redirect. Maybe.  Just telling you in case it gave you any ideas.

**Comment by creecros** [comment]
Oh,maybe add a log section that reads the logs and displays relevant logs entries related to each plugin....

**Reply by aljawaid** [reply]
hmmmm.... that leads to another plugin if youre willing to collaborate and create the functions for it... a log/error viewer to output all apache/system/kanboard error logs related to kanboard.... dont know if thats even possible... my log files are created daily based on domain... i dunno if other systems do the same
